### PR TITLE
Feature/persistant buttons

### PR DESCRIPTION
**Pull request by weeryan17**
# Pull request

 - [x] I have read and agreed to the [code of conduct](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CODE_OF_CONDUCT.md).
 - [x] I have read and complied with the [contributing guidelines](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CONTRIBUTING.md).
 - [x] What I'm implementing was assigned to me or was not being worked on and is a somewhat requested feature. For reference see our [GitHub projects](https://github.com/orgs/CascadeBot/projects).
 - [x] I have tested all of my changes.
 
## Added/Changed
Added extension to the button group for persistent buttons.
stored persistent buttons in guild data.
Moved todo and votes over to persistent.
Made messaging automatically store buttons in the correct places

### CODE REVIEW

File: src/main/java/org/cascadebot/cascadebot/UnicodeConstants.java [hunk 1]
```diff
@@ -48,6 +48,33 @@
...
+    public static final String Q = "\uD83C\uDDF6";
+    public static final String R = "\uD83C\uDDF7";
+    public static final String S = "\uD83C\uDDF8";
+    public static final String T = "\uD83C\uDDF9";
+    public static final String U = "\uD83C\uDDFA";
+    public static final String V = "\uD83C\uDDFB";
+    public static final String W = "\uD83C\uDDFC";
+    public static final String X = "\uD83C\uDDFD";
+    public static final String Y = "\uD83C\uDDFE";
+    public static final String Z = "\uD83C\uDDFF";
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/UnicodeConstants.java]: Following the standard in UnicodeConstants, we should probably have a comment with the emoji

File: src/main/java/org/cascadebot/cascadebot/events/ButtonEventListener.java [hunk 2]
```diff
@@ -33,15 +36,25 @@ public void onMessageReactionAdd(MessageReactionAddEvent e) {
...
+                if(data.getPersistentButtons().get(channel.getIdLong()).containsKey(e.getMessageIdLong())) {
+                    ButtonGroup group = data.getPersistentButtons().get(channel.getIdLong()).get(e.getMessageIdLong());
+                    doButtonPress(group, e.getReaction(), e.getMember(), e.getTextChannel());
                 }
             }
         }
     }
 
+    private void doButtonPress(ButtonGroup group, MessageReaction reaction, Member sender, TextChannel channel) {
+        Metrics.INS.buttonsPressed.labels(reaction.getReactionEmote().getName()).inc();
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/events/ButtonEventListener.java]: Maybe we could add a label about what type of button this is? i.e. persistent or not

File: src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java [hunk 3]
```diff
@@ -14,14 +15,17 @@
 import net.dv8tion.jda.api.entities.TextChannel;
 import org.cascadebot.cascadebot.CascadeBot;
 
+import javax.naming.OperationNotSupportedException;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java]: Random import?

File: src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java [hunk 4]
```diff
@@ -90,7 +95,7 @@ public void setMessage(long id) {
     }
 
     public void handleButton(Member clicker, TextChannel channel, Message buttonMessage, MessageReaction.ReactionEmote emote) {
-        for (Button button : buttons) {
+        for (Button button : this.getButtons()) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java]: Why the change? Doesn't make a difference just wondering why

File: src/main/java/org/cascadebot/cascadebot/utils/buttons/PersistentButtonGroup.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    public List<Button> getButtons() {
+        return persistentButtons.stream().map(PersistentButton::getButton).collect(Collectors.toList());
+    }
+
+    @Override
+    public void addButton(Button button) {
+    }
+
+    @Override
+    public void removeButton(Button button) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/utils/buttons/PersistentButtonGroup.java]: If these are not supposed to be used, they should probably throw an exception inside to make debugging easier. Something like `UnsupportedOperationException` would work perfect in this case

File: src/main/java/org/cascadebot/cascadebot/utils/votes/VoteMessageType.java [hunk 6]
```diff
@@ -20,11 +20,6 @@
     /**
      * This vote type will use discord regional indicator letters up to the amount specified in the builder.
      */
-    LETTERS,
-
-    /**
-     * This vote type will give you complete control over what reactions you want to use .
-     */
-    CUSTOM
+    LETTERS
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/utils/votes/VoteMessageType.java]: Where's the custom type gone?

File: src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java (hunk 4 above)
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java]: It actually does make a difference. This allows it to get the correct list of buttons for say persistent buttons. Persistent buttons actually doesn't use the standard button list at all.

File: src/main/java/org/cascadebot/cascadebot/utils/buttons/PersistentButtonGroup.java (hunk 5 above)
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/utils/buttons/PersistentButtonGroup.java]: I tried but it wanted me to add a throw clause to the method and therefor the parent method which broke everything.

File: src/main/java/org/cascadebot/cascadebot/utils/votes/VoteMessageType.java (hunk 6 above)
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/utils/votes/VoteMessageType.java]: Because the custom type is hard to do now that they're persistent the way it was being done before. At some point I might return to this.

File: src/main/java/org/cascadebot/cascadebot/utils/buttons/PersistentButtonGroup.java (hunk 5 above)
Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/utils/buttons/PersistentButtonGroup.java]: If you use `java.lang.UnsupportedOperationException` it won't force you to try catch since it's a runtime exception. I think before you tried using the javax one which is probably a checked exception

Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/utils/buttons/PersistentButtonGroup.java]: I'll try that

File: src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java (hunk 4 above)
Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java]: Oh I see, this is used so it gets the button list from the overrided method instead. Got it 👌

File: src/main/java/org/cascadebot/cascadebot/utils/votes/VoteButtonGroupBuilder.java [hunk 7]
```diff
@@ -57,12 +58,12 @@ public VoteButtonGroupBuilder(VoteMessageType type) {
...
     public VoteButtonGroupBuilder setOptionsAmount(int amount) {
-        if (type == VoteMessageType.YES_NO || type == VoteMessageType.CUSTOM) {
-            throw new UnsupportedOperationException("Cannot set options amount for yes no votes, or custom votes");
+        if (type == VoteMessageType.YES_NO) {
+            throw new UnsupportedOperationException("Cannot set options amount for yes no votes");
         }
 
-        if (type == VoteMessageType.LETTERS && amount > 26) {
-            throw new UnsupportedOperationException("Cannot have more then 26 options when using letters");
+        if (type == VoteMessageType.LETTERS && amount > 15) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/utils/votes/VoteButtonGroupBuilder.java]: Could you just leave a note about why this is 15? Just for future reference (We will forget)

File: src/main/java/org/cascadebot/cascadebot/data/database/DatabaseManager.java [hunk 8]
```diff
@@ -37,7 +37,8 @@
             CodecRegistries.fromProviders(PojoCodecProvider.builder().register(
                     "org.cascadebot.cascadebot.data.objects",
                     "org.cascadebot.cascadebot.permissions.objects",
-                    "org.cascadebot.shared"
+                    "org.cascadebot.shared",
+                    "org.cascadebot.cascadebot.utils.buttons"
```
[the remark this answers is not in the document]
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/data/database/DatabaseManager.java]: 🤦

File: src/main/java/org/cascadebot/cascadebot/data/objects/GuildData.java [hunk 9]
```diff
@@ -22,11 +22,14 @@
...
 import org.cascadebot.cascadebot.utils.buttons.ButtonGroup;
 import org.cascadebot.cascadebot.utils.buttons.ButtonsCache;
+import org.cascadebot.cascadebot.utils.buttons.PersistentButtonGroup;
 import org.cascadebot.cascadebot.utils.pagination.PageCache;
 
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/data/objects/GuildData.java]: Unused import :P
